oap-server/server-storage-plugin/storage-jdbc-hikaricp-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/jdbc/h2/dao/H2LogQueryDAO.java

```diff
@@ -91,14 +86,20 @@ public Logs queryLogs(String serviceId,
                                                        .provider()
                                                        .getService(ConfigService.class);
             searchableTagKeys = Arrays.asList(configService.getSearchableLogsTags().split(Const.COMMA));
-            if (searchableTagKeys.size() > maxSizeOfArrayColumn) {
-                searchableTagKeys = searchableTagKeys.subList(0, maxSizeOfArrayColumn);
-            }
         }
         StringBuilder sql = new StringBuilder();
         List<Object> parameters = new ArrayList<>(10);
 
-        sql.append("from ").append(LogRecord.INDEX_NAME).append(" where ");
+        sql.append("from ").append(LogRecord.INDEX_NAME);
+        if (!CollectionUtils.isEmpty(tags)) {
+            for (int i = 0; i < tags.size(); i++) {
+                sql.append(" inner join ").append(AbstractLogRecord.ADDITIONAL_TABLE_TAG).append(" ");
+                sql.append(AbstractLogRecord.ADDITIONAL_TABLE_TAG + i);
+                sql.append(" on ").append(LogRecord.INDEX_NAME).append(".").append(ID_COLUMN).append(" = ");
+                sql.append(AbstractLogRecord.ADDITIONAL_TABLE_TAG + i).append(".").append(ID_COLUMN);
+            }
```

Comment:
> This would join the tables for `tags.size()` times, it might take considerable time, can we use `inner join xxx on key1=? or key2=?` to only join the tables once ?

I am afraid this is not the correct logic. We discussed this many times for this in these 2 days, 2 tags actually mean, this segment should have both tags, not `one-of`.